### PR TITLE
wire/internal/wire: use on-disk GOPATH in generate tests

### PR DESCRIPTION
**Pull request by zombiezen**
The primary motivation is to permit a move to using `go/packages` instead of `go/loader`. `go/packages` runs exclusively by shelling out to the go tool, which precludes use of the in-memory "magic" GOPATH being used up to this point.

This has a secondary effect of removing a lot of code to support "magic" GOPATH from the test infrastructure. This is on the whole good, but necessitated a change in the error scrubbing: since the filenames are no longer fixed, error scrubbing also must remove the leading `$GOPATH/src` lines.

Another related change: since all callers of `Generate` needed to know the package path in order to write out wire_gen.go (necessitating a find-only import search) and `Generate` already has this information, `Generate` now returns this information to the caller. This should further reduce callers' coupling to Wire's load internals. It also eliminates code duplication.

This should hopefully shake out any difference in path separators for running on Windows, but I have not tested that yet.

Updates #78
Updates #323

### CODE REVIEW

File: wire/internal/wire/parse.go [hunk 1]
```diff
@@ -193,8 +194,14 @@ type Value struct {
 // Load finds all the provider sets in the given packages, as well as
```
Comment (vangent) [inline wire/internal/wire/parse.go]: Can you clarify the relationship between "given packages" and `patterns`? Same in `load`.

File: wire/internal/wire/wire.go [hunk 2]
```diff
@@ -36,32 +38,71 @@ import (
 	"golang.org/x/tools/go/loader"
 )
 
+// A GeneratedFile stores the content of a call to Generate and the
```
Comment (vangent) [inline wire/internal/wire/wire.go]: Nit: drop the "A ".

File: wire/internal/wire/wire_test.go [hunk 3]
```diff
@@ -332,31 +325,91 @@ func TestDisambiguate(t *testing.T) {
...
-				panic("BREAK1")
-			}
 			return false
 		}
 		i += sz
 	}
 	return true
 }
 
+func scrubError(gopath string, s string) string {
```
Comment (vangent) [inline wire/internal/wire/wire_test.go]: It's non-trivial to figure out what's being scrubbed. Can you add a docstring explaining, maybe with an example? Is this not possible to do with a regex?

File: wire/internal/wire/parse.go (hunk 1 above)
Comment (zombiezen) [reply wire/internal/wire/parse.go]: Done.

File: wire/internal/wire/wire.go (hunk 2 above)
Comment (zombiezen) [reply wire/internal/wire/wire.go]: Done.

File: wire/internal/wire/wire_test.go (hunk 3 above)
Comment (zombiezen) [reply wire/internal/wire/wire_test.go]: It is difficult to do with regex, as the initial match must be a literal string. Also, I will cite [Rob Pike's advice](https://commandcenter.blogspot.com/2011/08/regular-expressions-in-lexing-and.html) on the subject of regexes in parsing.

I've added a doc comment.

File: wire/internal/wire/parse.go [hunk 4]
```diff
@@ -191,9 +191,9 @@ type Value struct {
 	info *types.Info
 }
 
-// Load finds all the provider sets in the given packages, as well as
-// the provider sets' transitive dependencies. It may return both errors
-// and Info.
+// Load finds all the provider sets in the packages that match the given
```
Comment (vangent) [inline wire/internal/wire/parse.go]: Can you define "match"? Is pattern a regexp, or is it a partial match, or a full string match, or ...?

Comment (zombiezen) [reply wire/internal/wire/parse.go]: There is currently no agreed upon definition of this pattern: it is (will be) at the whims of what [`go/packages`](https://godoc.org/golang.org/x/tools/go/packages#pkg-overview) documents. How do you wish to proceed?

Comment (vangent) [reply wire/internal/wire/parse.go]: I would suggest just saying that. E.g., "... that match the given patterns according to the match rules defined by https://godoc.org/golang.org/x/tools/go/packages#pkg-overview..."

Comment (zombiezen) [reply wire/internal/wire/parse.go]: Done.

**Pull request merged**